internal/args/unmarshal.go

```diff
@@ -190,70 +185,95 @@ func set(dest reflect.Value, keys []string, value string) error {
 
 	switch dest.Kind() {
 	case reflect.Ptr:
-		// If type is a pointer we create a new Value and call set with the pointer.Elem()
-		newValue := reflect.New(dest.Type().Elem())
-		dest.Set(newValue)
-		return set(newValue.Elem(), keys, value)
+		// If type is a nil pointer we create a new Value. NB: maps and slices are pointers.
+		if dest.IsNil() {
+			dest.Set(reflect.New(dest.Type().Elem()))
+		}
+
+		// Call set with the pointer.Elem()
+		return set(dest.Elem(), argNameWords, value)
 
 	case reflect.Slice:
 		// If type is a slice:
-		// We check if keys[0] is an number to handle cases like keys.0.value=12
-		isIndex := regexp.MustCompile("^\\d+$")
-		if len(keys) > 0 && isIndex.MatchString(keys[0]) {
-			index, _ := strconv.Atoi(keys[0])
-			// If key is an number we make sure array is big enough to access the correct index.
-			if index >= dest.Len() {
-				diffLen := index - dest.Len() + 1
-				// To make the slice bigger we create a fake slice of diffLen size and we append it.
-				dest.Set(reflect.AppendSlice(dest, reflect.MakeSlice(dest.Type(), diffLen, diffLen)))
-			}
-
-			// Recursively call set without the index key
-			return set(dest.Index(index), keys[1:], value)
+		// We check if argNameWords[0] is an number to handle cases like keys.0.value=12
+
+		// We cannot handle slice without an index notation.
+		if len(argNameWords) == 0 {
+			return fmt.Errorf("missing index on the array")
 		}
 
-		// We can't handle nested message in a slice without index notation.
-		if len(keys) > 0 {
-			return fmt.Errorf("cannot handle nested struct without a slice index")
+		// Make sure index is a positive integer.
+		index, err := strconv.ParseUint(argNameWords[0], 10, 64)
+		if err != nil {
+			return fmt.Errorf("invalid index: '%s' is not a positive integer", argNameWords[0])
 		}
 
-		// We create a new value call set and append it to the slice
-		newValue := reflect.New(dest.Type().Elem())
-		err := set(newValue.Elem(), keys, value)
-		dest.Set(reflect.Append(dest, newValue.Elem()))
-		return err
+		// Make sure array is big enough to access the correct index.
+		diff := int(index) - dest.Len()
+		switch {
+		case diff > 2:
+			return fmt.Errorf("missing indices in the array: trying to set array at index %d before indices %v", index, missingIndices(int(index), dest.Len()))
+		case diff == 1:
```

Comment:
could be handle in a single case with the one before

Reply:
When we'll switch to a typed struct this will make a lot of sense to do this logic in the `Error()` func